source/sink/service_impl.cc

```diff
@@ -0,0 +1,132 @@
+  return grpc_status;
+}
+
+grpc::Status SinkServiceImpl::SinkRequestStream(
+    grpc::ServerContext*,
+    grpc::ServerReaderWriter<nighthawk::SinkResponse, nighthawk::SinkRequest>* stream) {
+  nighthawk::SinkRequest request;
+  while (stream->Read(&request)) {
+    ENVOY_LOG(trace, "Inbound SinkRequest {}", request.DebugString());
+    absl::StatusOr<std::vector<nighthawk::client::ExecutionResponse>> execution_responses =
+        sink_->LoadExecutionResult(request.execution_id());
+    if (!execution_responses.status().ok()) {
+      return abslStatusToGrpcStatus(execution_responses.status());
+    }
+    absl::StatusOr<nighthawk::client::ExecutionResponse> response =
+        mergeExecutionResponses(request.execution_id(), *execution_responses);
+    if (!response.status().ok()) {
+      return abslStatusToGrpcStatus(response.status());
+    }
+    nighthawk::SinkResponse sink_response;
+    *(sink_response.mutable_execution_response()) = *response;
+    if (!stream->Write(sink_response)) {
+      return abslStatusToGrpcStatus(
+          absl::Status(absl::StatusCode::kInternal, "Failure writing response to stream."));
+    }
+  }
+  return abslStatusToGrpcStatus(absl::OkStatus());
+}
+
+absl::Status mergeOutput(const nighthawk::client::Output& input_to_merge,
+                         nighthawk::client::Output& merge_target) {
+  if (!merge_target.has_options()) {
+    // If no options are set, that means this is the first part of the merge.
+    // Set some properties that shouldbe equal amongst all Output instances.
+    *merge_target.mutable_options() = input_to_merge.options();
+    *merge_target.mutable_timestamp() = input_to_merge.timestamp();
+    *merge_target.mutable_version() = input_to_merge.version();
+  } else {
+    // Options used should not diverge for a executions under a single execution id.
+    // Versions probably shouldn't either. We sanity check these things here, and
+    // report on error when we detect any mismatch.
+    if (!MessageDifferencer::Equivalent(input_to_merge.options(), merge_target.options())) {
+      return absl::Status(absl::StatusCode::kInternal,
+                          fmt::format("Options divergence detected: {} vs {}.",
+                                      merge_target.options().DebugString(),
+                                      input_to_merge.options().DebugString()));
+    }
+    if (!MessageDifferencer::Equivalent(input_to_merge.version(), merge_target.version())) {
+      return absl::Status(absl::StatusCode::kInternal,
+                          fmt::format("Version divergence detected: {} vs {}.",
+                                      merge_target.version().DebugString(),
+                                      input_to_merge.version().DebugString()));
+    }
+  }
+  // Append all input results into our own results.
+  for (const nighthawk::client::Result& result : input_to_merge.results()) {
+    merge_target.add_results()->MergeFrom(result);
+  }
```

Comment:
In addition to the above, I'm concerned about this more architecturally.  I would have thought the purpose of the full Execution Response would be to allow aggregation of all of the responses into a single cross-nighthawk global result object.

As it is, isn't this now a massive memory risk for the Sink service?  Will it run out of memory holding onto all of these execution responses.

Without the ability to combine all results into a single global result, it seems to me that this would be safer if we instead provided a stream of execution responses back.  That way, we wouldn't have to hold onto the memory for all of these different (potentially large) protos at the same time.

Reply:
Before we act on any of this, I want to get @mum4k 's eyes on it.  Because he's been more involved with you on this design, and we don't want to expand scope at all.

Reply:
I can see the concern here; however in the current state there are no "raw" statistics being transferred,
just the summaries (one for each worker thead involved + a summary per process/load-gen-service).

The raw statistics have been what I was thinking about mostly when being concerned about memory. 
That will be addressed transparantly to this service when we introduce it, but I can see that is hard to assess by just looking at this PR: I have tried to intentionally build up the functionality to keep review more fun and focussed.
We don't tag on an aggregated summary across all load gen services yet, and as such we don't have the raw stats being used yet.

Having said that: if we should anticipate volumes of results to an extent that this is a problem, then your concern is very valid, and my assumption here is off. In that case I'd be happy to iterate. I think that is solvable in various ways without too much effort.

Lastly, if we can, I do have a preference: I think that it is useful to propagate the information at this level by default. The summaries can sometimes be misleading -- at least the way we do them right now. A practical example I've seen in the wild: one node running a cpu hog (prometheus), significantly impacting compute capacity of a load generator running on the same node. It's not a silver bullet, but being able to look at all the details improves chances of catching hints of such things.